FEA Add support for float32 on `PairwiseDistancesReduction` using Tempita

jjerphan opened the pull request:
#### Reference Issues/PRs

Follows-up https://github.com/scikit-learn/scikit-learn/pull/22134

#### What does this implement/fix? Explain your changes.

This ports `PairwiseDistancesReduction` and other implementations to 32bit using Tempita.

#### Benchmarks results

The hardware scalability plateaus at 64 threads because, asymptotically and using Adham's law, 2.5% of the code (which parts of it are due to interaction with CPython) is sequential. 

Improved hardware scalability beyond that point, mean removing the last portions of sequential code accounting for the few points of percentage left.

![1](https://user-images.githubusercontent.com/13029839/178106120-bdafc400-ece7-4d5a-8a74-24db8034462d.png)

<details>
<summary>Raw results</summary>

```
    n_threads  n_train  n_test  n_features  mean_runtime  stderr_runtime
0           1   100000  100000          50     54.270973               0
1           2   100000  100000          50     27.357690               0
2           4   100000  100000          50     13.772927               0
3           8   100000  100000          50      7.034176               0
4          16   100000  100000          50      3.851457               0
5          32   100000  100000          50      2.134666               0
6          64   100000  100000          50      1.482027               0
7         128   100000  100000          50      2.239688               0
8           1   100000  100000         100     77.925089               0
9           2   100000  100000         100     39.125349               0
10          4   100000  100000         100     19.810733               0
11          8   100000  100000         100     10.130284               0
12         16   100000  100000         100      5.506694               0
13         32   100000  100000         100      3.067685               0
14         64   100000  100000         100      2.061337               0
15        128   100000  100000         100      3.396916               0
16          1   100000  100000         500    274.099079               0
17          2   100000  100000         500    138.078319               0
18          4   100000  100000         500     70.136737               0
19          8   100000  100000         500     35.598209               0
20         16   100000  100000         500     19.321611               0
21         32   100000  100000         500     10.415704               0
22         64   100000  100000         500      7.194686               0
23        128   100000  100000         500     12.095341               0
```

</details>

![2](https://user-images.githubusercontent.com/13029839/178106171-e23c4652-dc21-4afe-a472-7a8472d3207a.png)

<details>
<summary>Details</summary>

```
    n_threads  n_train  n_test  n_features  mean_runtime  stderr_runtime
0           1  1000000   10000          50     53.670283               0
1           2  1000000   10000          50     27.603052               0
2           4  1000000   10000          50     14.014151               0
3           8  1000000   10000          50      7.138670               0
4          16  1000000   10000          50      3.810226               0
5          32  1000000   10000          50      2.129321               0
6          64  1000000   10000          50      1.363076               0
7         128  1000000   10000          50      1.540974               0
8           1  1000000   10000         100     77.725753               0
9           2  1000000   10000         100     39.835435               0
10          4  1000000   10000         100     20.107033               0
11          8  1000000   10000         100     10.242633               0
12         16  1000000   10000         100      5.499012               0
13         32  1000000   10000         100      3.151450               0
14         64  1000000   10000         100      2.051802               0
15        128  1000000   10000         100      2.319589               0
16          1  1000000   10000         500    274.992947               0
17          2  1000000   10000         500    140.689740               0
18          4  1000000   10000         500     70.843511               0
19          8  1000000   10000         500     36.023845               0
20         16  1000000   10000         500     19.761463               0
21         32  1000000   10000         500     10.633548               0
22         64  1000000   10000         500      7.017808               0
23        128  1000000   10000         500      8.313477               0
```

</details>

### Benchmarks results between `main` (a5d50cf3c7) and this PR @ 31b8b28bcb969ef8c43f3f41cc9f216e2d73c764 (via 2c842bd9fac745cdbf98be6ac868553ace06d7ac)

Between ×1.2 and, well, ×250+ speed-ups: it looks like it just scales linearly.
Regressions are due to using too many cores when the size of the problem (i.e. `n_train` and `n_test`) is small.

<details>
<summary>1 thread</summary>

```· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         2/9 failed
[ 50.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     11.3±0.2ms     105±2ms       1.06±0s    
                10000     89.6±0.9ms     878±4ms      8.76±0.01s  
               10000000    1.41±0m        failed        failed    
              ========== ============ ============= ==============

[ 50.00%] ···· For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     20.1±0.1ms     193±1ms       2.02±0s    
                10000      203±1ms       2.12±0s       21.0±0s    
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      20.1±0.1ms       11.3±0.2ms     0.56  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-         193±1ms          105±2ms     0.54  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-         2.02±0s          1.06±0s     0.52  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)
-         203±1ms       89.6±0.9ms     0.44  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         21.0±0s       8.76±0.01s     0.42  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 100000, 100)
-         2.12±0s          878±4ms     0.41  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>


<details>
<summary>2 threads</summary>

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         2/9 failed
[ 50.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     6.89±0.2ms    53.9±0.7ms     531±3ms    
                10000     46.7±0.2ms    443±0.5ms      4.33±0s    
               10000000   42.7±0.06s      failed        failed    
              ========== ============ ============= ==============

[ 50.00%] ···· For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     18.3±0.2ms     171±1ms       1.78±0s    
                10000     179±0.9ms      1.86±0s       18.6±0s    
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      18.3±0.2ms       6.89±0.2ms     0.38  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-         171±1ms       53.9±0.7ms     0.32  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-         1.78±0s          531±3ms     0.30  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)
-       179±0.9ms       46.7±0.2ms     0.26  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         1.86±0s        443±0.5ms     0.24  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
-         18.6±0s          4.33±0s     0.23  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 100000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

<details>
<summary>4 threads</summary>

```· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         2/9 failed
[ 50.00%] ··· ========== ============= ============= ==============
              --                    n_test / n_features            
              ---------- ------------------------------------------
               n_train     1000 / 100   10000 / 100   100000 / 100 
              ========== ============= ============= ==============
                 1000     5.60±0.07ms    29.6±0.4ms     276±2ms    
                10000      27.3±0.3ms     230±1ms      2.23±0.01s  
               10000000    21.7±0.01s      failed        failed    
              ========== ============= ============= ==============

[ 50.00%] ···· For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     17.4±0.2ms    163±0.8ms      1.68±0s    
                10000      172±1ms       1.77±0s      17.5±0.02s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      17.4±0.2ms      5.60±0.07ms     0.32  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-       163±0.8ms       29.6±0.4ms     0.18  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-         1.68±0s          276±2ms     0.16  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)
-         172±1ms       27.3±0.3ms     0.16  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         1.77±0s          230±1ms     0.13  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

<details>
<summary>8 threads</summary>

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[ 50.00%] ··· ========== ============= ============= ==============
              --                    n_test / n_features            
              ---------- ------------------------------------------
               n_train     1000 / 100   10000 / 100   100000 / 100 
              ========== ============= ============= ==============
                 1000     6.83±0.03ms    17.1±0.3ms     150±2ms    
                10000      17.3±0.3ms     122±1ms      1.16±0.01s  
               10000000    11.5±0.01s     1.89±0m        failed    
              ========== ============= ============= ==============

[ 50.00%] ···· For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     18.8±0.2ms    167±0.7ms      1.71±0s    
                10000     176±0.6ms      1.80±0s      17.7±0.01s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      18.8±0.2ms      6.83±0.03ms     0.36  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-       167±0.7ms       17.1±0.3ms     0.10  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-       176±0.6ms       17.3±0.3ms     0.10  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         1.71±0s          150±2ms     0.09  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)
-         1.80±0s          122±1ms     0.07  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>


<details>
<summary>16 threads</summary>

```· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[ 50.00%] ··· ========== ============= ============= ==============
              --                    n_test / n_features            
              ---------- ------------------------------------------
               n_train     1000 / 100   10000 / 100   100000 / 100 
              ========== ============= ============= ==============
                 1000     9.97±0.08ms    60.3±0.3ms    87.2±0.6ms  
                10000      15.7±0.2ms    106±0.6ms      631±3ms    
               10000000    6.22±0.02s    59.4±0.2s       failed    
              ========== ============= ============= ==============

[ 50.00%] ···· For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     20.5±0.3ms     168±1ms       1.67±0s    
                10000      175±1ms       1.82±0s      18.0±0.02s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      20.5±0.3ms      9.97±0.08ms     0.49  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-         168±1ms       60.3±0.3ms     0.36  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-         175±1ms       15.7±0.2ms     0.09  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         1.82±0s        106±0.6ms     0.06  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
-         1.67±0s       87.2±0.6ms     0.05  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

<details>
<summary>32 threads</summary>

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[ 50.00%] ··· ========== ============= ============= ==============
              --                    n_test / n_features            
              ---------- ------------------------------------------
               n_train     1000 / 100   10000 / 100   100000 / 100 
              ========== ============= ============= ==============
                 1000     17.4±0.08ms    95.7±0.3ms    59.3±0.3ms  
                10000      21.0±0.3ms    92.8±0.3ms     366±20ms   
               10000000    3.49±0.02s    32.3±0.1s       failed    
              ========== ============= ============= ==============

[ 50.00%] ···· For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     21.4±0.5ms     173±2ms       1.64±0s    
                10000      179±3ms       1.82±0s      18.1±0.01s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-      21.4±0.5ms      17.4±0.08ms     0.81  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-         173±2ms       95.7±0.3ms     0.55  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
-         179±3ms       21.0±0.3ms     0.12  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-         1.82±0s       92.8±0.3ms     0.05  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
-         1.64±0s       59.3±0.3ms     0.04  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

<details>
<summary>64 threads</summary>

```
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
·· Installing 31b8b28b <feat/pdr-32bit> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[ 50.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     33.2±10ms      185±20ms     49.0±0.2ms  
                10000     31.2±0.3ms     169±10ms      246±20ms   
               10000000   2.50±0.04s    20.0±0.02s      failed    
              ========== ============ ============= ==============

[ 50.00%] ···· For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000      37.9±5ms      200±2ms       1.76±0s    
                10000     208±0.8ms     2.07±0.01s    20.1±0.04s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
-       208±0.8ms       31.2±0.3ms     0.15  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-      2.07±0.01s         169±10ms     0.08  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
-         1.76±0s       49.0±0.2ms     0.03  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

<details>
<summary>128 threads</summary>

```
· Creating environments
· Discovering benchmarks
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 31b8b28b <feat/pdr-32bit> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[ 50.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000      250±30ms     1.45±0.1s     13.9±0.1s   
                10000      239±10ms     1.42±0.05s    12.8±0.1s   
               10000000   1.66±0.02s    13.1±0.08s      failed    
              ========== ============ ============= ==============

[ 50.00%] ···· For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         3/9 failed
[100.00%] ··· ========== ============ ============= ==============
              --                    n_test / n_features           
              ---------- -----------------------------------------
               n_train    1000 / 100   10000 / 100   100000 / 100 
              ========== ============ ============= ==============
                 1000     51.2±10ms      225±2ms      1.85±0.04s  
                10000      226±2ms      2.22±0.03s    21.3±0.03s  
               10000000     failed        failed        failed    
              ========== ============ ============= ==============

[100.00%] ···· For parameters: 10000000, 1000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 10000, 100
               
               
               asv: benchmark timed out (timeout 500s)
               
               For parameters: 10000000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [31b8b28b]
     <main>           <feat/pdr-32bit>
+         225±2ms        1.45±0.1s     6.45  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
+       51.2±10ms         250±30ms     4.89  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
-      2.22±0.03s       1.42±0.05s     0.64  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
-      21.3±0.03s        12.8±0.1s     0.60  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 100000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

</details>


##### Benchmarks information

<details>
<summary>Machine specification</summary>

```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              256
On-line CPU(s) list: 0-255
Thread(s) per core:  2
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7742 64-Core Processor
Stepping:            0
CPU MHz:             3388.360
BogoMIPS:            4491.59
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-63,128-191
NUMA node1 CPU(s):   64-127,192-255
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

```

</details>

jjerphan committed:
FEA Port implementations to 32bit using Tempita

jjerphan committed:
Validate Y_norm_squared

ogrisel commented:
Thanks for the updated PR. I assume that merging with main is needed before starting to review this.

Could you run the benchmarks with more imbalanced train / tests, e.g. `n_samples_train = int(1e7)` and `n_samples_test = 1000`?

I wonder if the performance slowdown for a very large number of threads is caused by the fact that we have two few chunks to execute per-thread and using more imbalanced benchmark cases might validate (or invalidate) this hypothesis.

ogrisel commented:
still, a 50x speed-up w.r.t. `main` is nice :)

jjerphan commented:
> Thanks for the updated PR. I assume that merging with main is needed before starting to review this.

You're welcome! I did not know if this PR was submitted correctly while travelling (🍀), another one is to come for a new `pairwise_distances` back-end.

> still, a 50x speed-up w.r.t. main is nice :)

Yes, I am quite glad we can reach those performance. I don't think we need to adapt the chunk size for the float32 case because there's little additional memory due to data-structures (the extra datastructures memory-wise are just the original `X_c` and `Y_c`) but this can be tried in another PR. :)

> Could you run the benchmarks with more imbalanced train / tests, e.g. n_samples_train = int(1e7) and n_samples_test = 1000?

Yes. Let's try that.

> I wonder if the performance slowdown for a very large number of threads is caused by the fact that we have two few chunks to execute per-thread and using more imbalanced benchmark cases might validate (or invalidate) this hypothesis.

I share the same hypothesis. I think we can explore a strategies to have a minimal number of batch per thread in another PR (a task I have added in the TODO list in the description).

jjerphan committed:
Merge branch 'main' into feat/pdr-32bit

jjerphan commented:
I've adapted the description with updated benchmarks script and results.

It looks like the implementations scales well on the `(n_samples_train, n_samples_test) = (int(1e7), 1000)` case. On `main`, the execution times out after 500s in all the cases, even when using 128 threads, but complete in less than 2s, reaching "×100+ speed-ups" in this case (see the raw logs).

The drop is mainly present when using too much threads. I think the `PairwiseDistancesArgKmin` used under the hood scales even better and that it is the sequential part at the beginning of `kneighbors` which might be costly.

Probably we could have had advertised the practical support for really large datasets more in the changelog for 1.1? :)

adrinjalali commented:
WOW

jjerphan committed:
MAINT Do not use PairwiseDistancesArgKmin for TSNE for now

jjerphan committed:
MAINT Adapt tests

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
+            with config_context(enable_cython_pairwise_dist=False):
+                # Distances have really tiny difference if using the
+                # PairwiseDistancesReductions on 32bit data.
+                # This causes in turn the jpint probability distributions computations.
```
```suggestion
                # This causes in turn the joint probability distributions computations.
```

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
+                # Distances have really tiny difference if using the
+                # PairwiseDistancesReductions on 32bit data.
+                # This causes in turn the jpint probability distributions computations.
+                # bellow to have significant differences and thus different results.
```
```suggestion
                # below to have significant differences and thus different results.
```

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
             t0 = time()
-            distances_nn = knn.kneighbors_graph(mode="distance")
+            with config_context(enable_cython_pairwise_dist=False):
+                # Distances have really tiny difference if using the
```
What is causing the difference?

thomasjpfan reviewed (commented):
We are on the path to make everything use Tempita :)

jjerphan committed:
DOC Fix typo

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
DOC Fix typo

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan commented:
> We are on the path to make everything use Tempita :)

It looks like yes. IMO, even if it's suboptimal, restrictive and hard to maintain on the long run, it's rather a pragmatic solution given where we are at today.

From IRL discussions this week :four_leaf_clover:, it looks like @adrinjalali is interested to experiment with alternatives, like Rust. I think it's worth exploring, but might add complexity especially on the build setup and on interfacing with other libraries like BLAS and colleagues.
Similar concerns also apply for C++.

Probably the work on https://github.com/scikit-learn/scikit-learn/issues/22438 might help?

jjerphan committed:
Retrigger CI

jjerphan committed:
DOC Add whats_new entry

ogrisel commented on doc/whats_new/v1.1.rst:
```diff
@@ -213,7 +213,11 @@ Changelog
   For instance :class:`sklearn.neighbors.NearestNeighbors.kneighbors` and
   :class:`sklearn.neighbors.NearestNeighbors.radius_neighbors`
-  can respectively be up to ×20 and ×5 faster than previously.
+  can respectively be up to ×20 and ×5 faster than previously on laptop.
```
```suggestion
  can respectively be up to ×20 and ×5 faster than previously on a laptop.
```

ogrisel commented on doc/whats_new/v1.2.rst:
```diff
@@ -71,6 +71,42 @@ Changelog
+
+  For instance :class:`sklearn.neighbors.NearestNeighbors.kneighbors` and
+  :class:`sklearn.neighbors.NearestNeighbors.radius_neighbors`
+  can respectively be up to ×20 and ×5 faster than previously on laptop.
```
```suggestion
  can respectively be up to ×20 and ×5 faster than previously on a laptop.
```

ogrisel commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
-            distances_nn = knn.kneighbors_graph(mode="distance")
+            with config_context(enable_cython_pairwise_dist=False):
+                # Distances have really tiny difference if using the
+                # PairwiseDistancesReductions on 32bit data.
```
```suggestion
                # PairwiseDistancesReductions on float32 data.
```

ogrisel commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
+            with config_context(enable_cython_pairwise_dist=False):
+                # Distances have really tiny difference if using the
+                # PairwiseDistancesReductions on 32bit data.
+                # This causes in turn the joint probability distributions computations.
```
```suggestion
                # This causes in turn the joint probability distributions computation
```

ogrisel commented on sklearn/manifold/_t_sne.py:
```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
+                # This causes in turn the joint probability distributions computations.
+                # below to have significant differences and thus different results.
+                # For backward compatibility on results, we prefer to use the
+                # old back-end.
```
I don't understand this comment. If the problem is about `float32` vs `float64`, we could enable the old numpy-based backend only for `float32` data and still use the `float64` for the new backend, no? Otherwise we will introduce a performance regression for `float64` data in 1.2 with respect to 1.1.

Do you understand why would the old backend be more numerically accurate than the new backend?

Which T-SNE tests in particular would fail if we were to use the new backend here?

jjerphan replied:
Yes, this is a really coarse restriction that I will refine.

---

If we remove the context, this test fails:
https://github.com/scikit-learn/scikit-learn/blob/d2f6a66c1740abe477d5bdffed62aa6e8e345fc6/sklearn/manifold/tests/test_t_sne.py#L1110-L1135

It does not fail on `main` because `sklearn.metrics.manhattan_distances` is used on both variant under the hood.

On this PR, distances get computed via `ManhattanDistance` in the `PairwiseDistancesReduction` infrastructure and there are small differences when using float32 for the default `method`.

Namely, when `method="barnes-hut"` for TNSE, `kneighbors_graph` which ultimately uses `PairwiseDistancesArgkmin` via `kneighbors` is called:

https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L933-L962

the max of the abs of the element-wise difference is `10e-7` for the distance matrix (`A_data`): 

https://github.com/scikit-learn/scikit-learn/blob/bcefbe1e3a5bc5ff112e4b2caf748e0d1fa6a304/sklearn/neighbors/_base.py#L989

Those small numerical differences are then propagated in those computations:
https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L974-L982

causing `P` to be slight different (`max(abs(diff_P)) = 10e-10`) and the gradient descent to give different results (`max(abs(diff_params)) = 18.360132` in this case):

https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L1067-L1075

---

ogrisel replied:
Ok thanks for the investigation. So those the root cause are expected machine precision level rounding errors.

This problem would disappear if we changed `manhattan_distances` to also use the new infrastructure, right?

Let's change this test to explain this and call `pytest.xfail`. I think in a follow-up PR we might want to refactor `sklearn.metrics.manhattan_distances` to rely on the new code which will make the problem go away, but let's do one thing at a time.

jjerphan replied:
> This problem would disappear if we changed manhattan_distances to also use the new infrastructure, right?

Yes. More generally, I think it would help to clarify the reliance on `scipy.spatial.distance` for scikit-learn.

jjerphan replied:
@thomasjpfan: FYI, here is a follow-up of discussions regarding failure for t-SNE.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_base.pxd.tp:
```diff
@@ -1,23 +1,40 @@
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

ogrisel replied:
Actually, wouldn't that be a good opportunity to stop using the `DTYPE(_t)` names that are quite confusing? Better be explicit.

jjerphan replied:
I agree that those aren't proper and understandable names, yet I would rather have other PRs unify custom types definitions and names. What do you think?

jjerphan commented on sklearn/metrics/_dist_metrics.pxd.tp:
```diff
@@ -5,9 +5,10 @@ implementation_specific_values = [
     # On the first hand, an empty string is used for `name_suffix`
-    # for the float64 case as to still be able to expose the original
-    # float64 implementation under the same API, namely `DistanceMetric`.
+    # for the 64bit case as to still be able to expose the original
+    # 64bit implementation under the same API, namely `DistanceMetric`.
```
```suggestion
    # for the float64 case as to still be able to expose the original
    # float64 implementation under the same API, namely `DistanceMetric`.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -1,17 +1,33 @@
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pxd.tp:
```diff
@@ -0,0 +1,41 @@
+    #
+    #       name_suffix, DistanceMetric, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -1,3 +1,18 @@
+    #
+    #       name_suffix, DistanceMetric, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pxd.tp:
```diff
@@ -1,11 +1,29 @@
+    #
+    #       name_suffix, need_upcast, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_radius_neighborhood.pxd.tp:
```diff
@@ -1,14 +1,24 @@
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/setup.py:
```diff
@@ -3,13 +3,30 @@
 import numpy as np
 from numpy.distutils.misc_util import Configuration
 
+from sklearn._build_utils import gen_from_templates
+
```
```suggestion
from sklearn._build_utils import gen_from_templates
```

jjerphan replied:
```suggestion
from sklearn._build_utils import gen_from_templates
```

jjerphan replied:
```suggestion
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pyx.tp:
```diff
@@ -1,3 +1,20 @@
+    #
+    #       name_suffix, need_upcast, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -1,17 +1,33 @@
 
 from sklearn import get_config
 from cython cimport final
+from cython.operator cimport dereference as deref
 from cython.parallel cimport parallel, prange
```
```suggestion
from cython.operator cimport dereference as deref
from cython.parallel cimport parallel, prange
from libc.stdlib cimport free, malloc
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -1,17 +1,33 @@
 
 from numbers import Integral
 from sklearn.utils import check_scalar
+from libc.stdlib cimport free, malloc
```
```suggestion
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp:
```diff
@@ -1,22 +1,26 @@
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pxd.tp:
```diff
@@ -0,0 +1,50 @@
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use 64bit types as defined in `sklearn.utils._typedefs`
+    # to maintain consistency.
```
```suggestion
    # We also use the float64 dtype and C-type names as defined in
    # `sklearn.utils._typedefs` to maintain consistency.
```

jjerphan committed:
DOC Rectify comments and perform some cosmetic changes

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/setup.py:
```diff
@@ -3,13 +3,29 @@
 import numpy as np
 from numpy.distutils.misc_util import Configuration
 
+from sklearn._build_utils import gen_from_templates
```
```suggestion
from sklearn._build_utils import gen_from_templates

```

jjerphan committed:
STY Make `black` happy

jjerphan committed:
MAINT Use the new back-end but  xfail on TSNE edge-case

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -15,6 +15,7 @@
 from scipy.spatial.distance import pdist
 from scipy.spatial.distance import squareform
 from scipy.sparse import csr_matrix, issparse
+
```
```suggestion
```

jjerphan commented on sklearn/metrics/_dist_metrics.pxd.tp:
```diff
@@ -5,6 +5,7 @@ implementation_specific_values = [
     #
     #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
     #
+    #
```
```suggestion
```

jjerphan committed:
STY Remove spurious changes

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -1107,32 +1107,45 @@ def test_gradient_bh_multithread_match_sequential():
     """Make sure that TSNE works for different distance metrics"""
+
+    if method == "barnes_hut" and metric == "manhattan":
+        # For reference, see:
+        # https://github.com/scikit-learn/scikit-learn/pull/23865/files#r925721573
+        pytest.xfail(
+            "Results are qualitatively different for method == 'barnes_hut' and metric"
+            " == 'manhattan', but qualitatively identical."
+        )
```
```suggestion
        # The distances computed by `manhattan_distances` differ slightly from those
        # computed internally by NearestNeighbors via the PairwiseDistancesReduction
        # Cython code-based. This in turns causes T-SNE to converge to a different
        # solution but this should not impact the qualitative results as both
        # methods.
        # NOTE: it's probably not valid from a mathematical point of view to use the
        # Manhattan distance for T-SNE...
        # TODO: re-enable this test if/when `manhattan_distances` is refactored to
        # reuse the same underlying Cython code NearestNeighbors.
        # For reference, see:
        # https://github.com/scikit-learn/scikit-learn/pull/23865/files#r925721573
        pytest.xfail(
            "Distance computations are different for method == 'barnes_hut' and metric"
            " == 'manhattan', but this is expected."
        )
```

jjerphan committed:
DOC Better motivate the xfail on t-SNE edge

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pxd.tp:
```diff
@@ -0,0 +1,50 @@
+    # We also use the float64 dtype and C-type names as defined in
+    # `sklearn.utils._typedefs` to maintain consistency.
+    #
+    ('64', 'DTYPE_t', 'DTYPE'),
```
**Not actionable**: I find the asymmetry with `cnp.float32_t` and `DTYPE_t` a bit weird. I prefer to have used `cnp.float64_t` everywhere and remove `DTYPE_t` all together. For someone new to the code base, it's easier to know what `cnp.float64_t` is at a glance. 

BTW, I would even go as far as using `float` and `double`. This is what NumPy typedefs `float*_t` to: [link1](https://github.com/numpy/numpy/blob/f1e50253a3b56375e795ee72032cca90a3832cd4/numpy/__init__.pxd#L285-L286) and [link2](https://github.com/numpy/numpy/blob/f1e50253a3b56375e795ee72032cca90a3832cd4/numpy/__init__.pxd#L703-L704). This way we do not even need to compile against NumPy.

That being said, I do not want to block this PR with this comment. I am okay with the `DTYPE_t` for now.

jjerphan replied:
> Not actionable: I find the asymmetry with cnp.float32_t and DTYPE_t a bit weird. I prefer to have used cnp.float64_t everywhere and remove DTYPE_t all together. For someone new to the code base, it's easier to know what cnp.float64_t is at a glance.

I used this for consistency with previous usage of Tempita in the code-base, but this is definitely hard to understand. I will use `cnp.float64_t` instead of `DTYPE_t` and `np.float64` instead of `DTYPE`.

> BTW, I would even go as far as using float and double. This is what NumPy typedefs float*_t to: link1 and link2. This way we do not even need to compile against NumPy.
> 
> That being said, I do not want to block this PR with this comment. I am okay with the DTYPE_t for now.

I agree that Cython implementations mixed a lot of a C-types and dtypes zoo, and would do follow-up PRs to unify it progressively across the whole code-base.

jjerphan replied:
> That being said, I do not want to block this PR with this comment. I am okay with the DTYPE_t for now.

Are you referring to the whole comment, or solely to its second part, namely

> BTW, I would even go as far as using float and double. This is what NumPy typedefs float*_t to: [link1](https://github.com/numpy/numpy/blob/f1e50253a3b56375e795ee72032cca90a3832cd4/numpy/__init__.pxd#L285-L286) and [link2](https://github.com/numpy/numpy/blob/f1e50253a3b56375e795ee72032cca90a3832cd4/numpy/__init__.pxd#L703-L704). This way we do not even need to compile against NumPy.

?

So as to know if I need to perform some renaming in this PR for the templating or if it would enlarge the scope of this PR too much.

thomasjpfan replied:
> if it would enlarge the scope of this PR too much.

It would increase the scope too much. I meant that the whole comment is not actionable in terms of this PR.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_base.pyx.tp:
```diff
@@ -44,12 +60,53 @@ cpdef DTYPE_t[::1] _sqeuclidean_row_norms64(
+
+    with nogil, parallel(num_threads=num_threads):
+        # Thread-local buffer allocation
+        X_i_upcast_ptr = <DTYPE_t* > malloc(sizeof(DTYPE_t) * d)
```
I do not think that this is much different, it's just that we need temporary buffers and memory management is relatively simple here.

I am +0 regarding using `std::vector` here, but probably using [`std::array`](https://en.cppreference.com/w/cpp/container/array) makes more sense. What do you think?

jjerphan replied:
I've pushed d91b8aca68ed4de1c70d4ddde5c10324e03ae7fc which uses `vector[vector[T]]` without `shared_ptr`.

jjerphan committed:
Use stack-allocated vector[vector[T]] instead of thread-local buffers

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pxd.tp:
```diff
@@ -16,6 +34,12 @@ cdef class GEMMTermComputer64:
         # Buffers for the `-2 * X_c @ Y_c.T` term computed via GEMM
         vector[vector[DTYPE_t]] dist_middle_terms_chunks
 
+{{if need_upcast}}
```
Very nit: What do you think of?

```suggestion
{{if name_suffix == "32"}}
```

It makes it more clear that 32bit requires upcasting

jjerphan replied:
That's a good remark. I think we can make clearer that the float32 implementations require upcasting whilst avoiding duplicating this logic. Also potentially in the future other dtypes might require upcast as well.

Hence, I would keep `need_upcast` but makes its definition clearer and more visible. What do you think? Furthermore, I think the naming can be improved. What about `upcast_to_float64` instead of `need_upcast`?

thomasjpfan replied:
If there are other dtypes that require upcasting, then having a boolean makes sense.

I do like `upcast_to_float64` over `need_upcast`.

jjerphan replied:
> If there are other dtypes that require upcasting, then having a boolean makes sense.

Just to make it explicit: at the moment, there's only float32 but other dtype might be supported in the future (though it's not planned). Is this OK for you?

thomasjpfan replied:
> Just to make it explicit: at the moment, there's only float32 but other dtype might be supported in the future (though it's not planned). Is this OK for you?

Yup. That's okay with me.

jjerphan replied:
Done via 2a826997bd21f2b2b8bcfaa92e163335555e5afa.

thomasjpfan reviewed (approved):
I left a minor nit: otherwise LGTM

jjerphan committed:
MAINT Reword "need_upcastr" to "upcast_to_float64"

Done with:

grep -rl need_upcast . | xargs sed -i's/need_upcast/upcast_to_float64/g'

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
MAINT Also ignore generated Cython source at the project level

jjerphan committed:
Merge branch 'main' into feat/pdr-32bit

jjerphan committed:
Merge branch 'main' into feat/pdr-32bit

jjerphan committed:
MAINT Fix dtype_validity

jjerphan committed:
Merge branch 'main' into feat/pdr-32bit

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pxd.tp:
```diff
@@ -0,0 +1,50 @@
+    # We also use the float64 dtype and C-type names as defined in
+    # `sklearn.utils._typedefs` to maintain consistency.
+    #
+    ('64', 'DTYPE_t', 'DTYPE'),
```
I would be in favor of stopping the use of the confusing `DTYPE_t` and `ITYPE_t` names in newly created source files.

```suggestion
    ('64','cnp.float64_t', 'np.float64'),
```

and delay the work to fix the existing Cython file names in other modules for later clean-up PRs (one per module to keep them manageable to review).

jjerphan replied:
I am OK only generating `cnp.float64_t` instead of `DTYPE_t` via Tempita for now and to then perform all the other changes in subsequent PRs, like #24153.

Before performing such change, I think maintainers first need to agree on the types to use for Cython's implementations: in https://github.com/scikit-learn/scikit-learn/pull/23865#discussion_r926716129, @thomasjpfan proposed to use concrete types (e.g. `float` and `double`) directly because numpy redefine them. This way we might be able to get ride of the Cython interfaces for numpy (i.e. `cnp`).

What do you think?

ogrisel replied:
Personnally I find the `cnp.float32_t` and `cnp.float64_t` names more explicit and their mapping to matching numpy dtype object more obvious for Cython new-comers.

No strong opinion though.

ogrisel replied:
I dont' like using "float" as a type name in discussions because one never knows if it refers to the floating-point type family as a whole (`dtype.kind == "f"`) or as to the specific 32-bit precision floating point representation.

jjerphan replied:
So do I, as it can also refer to python `float` which are `PyObject` wrapping C `double`.

I would use numpy types and dtypes for clarity, but I think we choose made the decision after discussions among maintainers working on Cython code. WDYT?

jjerphan committed:
MAINT Use types and dtypes via NumPy directly

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel reviewed (approved):
I did another quick pass. LGTM. Let's merge and handle type renaming in dedicated PRs (e.g.  #24153).

ogrisel commented:
:tada:

jjerphan commented:
Thanks @ogrisel and @thomasjpfan for the reviews!
